### PR TITLE
1.18-pre4 NBT

### PR DESCRIPTION
**Pull request by apple502j**
Can't believe Mojang made this in less than a week.

### CODE REVIEW

File: mappings/net/minecraft/nbt/NbtType.mapping [hunk 1]
```diff
@@ -15,3 +15,19 @@ CLASS net/minecraft/class_4614 net/minecraft/nbt/NbtType
...
+		ARG 1 input
+		ARG 2 count
+	METHOD method_39877 visitOrSkip (Ljava/io/DataInput;Lnet/minecraft/class_6836;)V
+		ARG 1 input
+		ARG 2 visitor
+	CLASS 1
+		METHOD method_39878 createException ()Ljava/io/IOException;
+	CLASS class_6839 SizedSkippable
+		METHOD method_39853 getSizeInBytes ()I
+	CLASS class_6840 Skippable
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtType.mapping]: This is more like
```suggestion
	CLASS class_6840 OfVariableSize
```

File: mappings/net/minecraft/nbt/NbtType.mapping [hunk 2]
```diff
@@ -15,3 +15,19 @@ CLASS net/minecraft/class_4614 net/minecraft/nbt/NbtType
...
+		ARG 2 visitor
+	METHOD method_39854 skip (Ljava/io/DataInput;I)V
+		ARG 1 input
+		ARG 2 count
+	METHOD method_39877 visitOrSkip (Ljava/io/DataInput;Lnet/minecraft/class_6836;)V
+		ARG 1 input
+		ARG 2 visitor
+	CLASS 1
+		METHOD method_39878 createException ()Ljava/io/IOException;
+	CLASS class_6839 SizedSkippable
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtType.mapping]:
```suggestion
	CLASS class_6839 OfFixedSize
```

File: mappings/net/minecraft/nbt/NbtType.mapping [hunk 3]
```diff
@@ -15,3 +15,19 @@ CLASS net/minecraft/class_4614 net/minecraft/nbt/NbtType
 		COMMENT after the NBT element is instantiated.
 		COMMENT
 		COMMENT @return {@code true} if this NBT type is immutable, else {@code false}
+	METHOD method_39851 readAndSkip (Ljava/io/DataInput;)V
+		ARG 1 input
+	METHOD method_39852 visit (Ljava/io/DataInput;Lnet/minecraft/class_6836;)Lnet/minecraft/class_6836$class_6838;
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtType.mapping]: Maybe
```suggestion
	METHOD method_39852 doAccept (Ljava/io/DataInput;Lnet/minecraft/class_6836;)Lnet/minecraft/class_6836$class_6838;
```

File: mappings/net/minecraft/nbt/NbtType.mapping [hunk 4]
```diff
@@ -15,3 +15,19 @@ CLASS net/minecraft/class_4614 net/minecraft/nbt/NbtType
...
+		ARG 1 input
+		ARG 2 visitor
+	METHOD method_39854 skip (Ljava/io/DataInput;I)V
+		ARG 1 input
+		ARG 2 count
+	METHOD method_39877 visitOrSkip (Ljava/io/DataInput;Lnet/minecraft/class_6836;)V
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtType.mapping]:
```suggestion
	METHOD method_39877 accept (Ljava/io/DataInput;Lnet/minecraft/class_6836;)V
```

File: mappings/net/minecraft/nbt/deserialize/EmptyVisitor.mapping [hunk 5]
```diff
@@ -0,0 +1,2 @@
+CLASS net/minecraft/class_6845 net/minecraft/nbt/deserialize/EmptyVisitor
```
Comment (liach) [inline mappings/net/minecraft/nbt/deserialize/EmptyVisitor.mapping]: This is a skipping visitor that skips by default. It is not empty as an implementation can choose to not skip certain elements.

**Review by liach — COMMENTED** [review]
A few problems. I should probably check it out locally and pr stuff to you

File: mappings/net/minecraft/nbt/deserialize/StackBasedVisitor.mapping [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+	METHOD method_39883 addToParent (Lnet/minecraft/class_2520;)V
+		ARG 1 nbt
+	METHOD method_39884 (Lnet/minecraft/class_2487;Lnet/minecraft/class_2520;)V
+		ARG 2 nbt
+	METHOD method_39885 (Lnet/minecraft/class_2520;)V
+		ARG 1 nbt
+	METHOD method_39886 pushStack (Lnet/minecraft/class_4614;)V
+		ARG 1 type
+	METHOD method_39887 getParent ()Lnet/minecraft/class_2520;
+	METHOD method_39888 getStackSize ()I
```
Comment (enbrain) [inline mappings/net/minecraft/nbt/deserialize/StackBasedVisitor.mapping]:
```suggestion
	METHOD method_39888 getDepth ()I
```

**Comment by liach** [comment]
> A few problems. I should probably check it out locally and pr stuff to you

See apple502j/yarn#1

**Comment by apple502j** [comment]
Thanks for the PR - yeah, this is just a temporary name I choose so that I can at least (sort of) understand what it's doing, without all of those confusing intermediary. I'll merge the PR and see if there's any better ones.

**Review by enbrain — APPROVED** [review]

**Review by liach — APPROVED** [review]

**Comment by liach** [comment]
I'm surprised they didn't add visit array size methods so the scanner can decide to skip array based on size (like for list meta) without having to parse array contents. i'd say this is more ad-hoc, somewhat like the tree api i wrote for mappings parser.

**Pull request merged**